src/libsystemd/sd-id128/id128-util.c

```diff
@@ -97,6 +97,15 @@ int id128_read_fd(int fd, Id128Format f, sd_id128_t *ret) {
 
         switch (l) {
 
+        case 13:
+        case 14:
+                /* Treat an "uninitialized" id file like an empty one */
+                if (strncmp(buffer, "uninitialized\n", l) == 0)
```

Comment:
> hmm, i don#t think we should do this unconditionally. It's too special. Let's enable this via some flag that can be ORed into the Id128Format type.

I'm not so sure whether this is a good idea.  I've implemented this to test how it would look and it doesn't feel right to me ... First of all, it would affect all in-tree places where `ID128_PLAIN` is currently used (short of the test-suite).  So the existence of `ID128_PLAIN` without the flag is only relevant to systemd API users at that point.

Secondly, I feel like this makes the mechanism a bit fragile.  People will easily forget this flag and have their code mistreat a machine-id from a system that crashed during first-boot.  As this is a case that will probably not see much test-coverage anyway, it could be hard to catch such incorrect code and nasty for those few who get to encounter it.  I myself wouldn't even have though about adding this extra case to `id128_read_fd()` if I didn't create a (local) test to try out an aborted boot (systemd-nspawn refusing to boot the image with "uninitialized" machine-id is how I found out).

As most code that reads `/etc/machine-id` will luckily use `id128_read()` or `id128_read_fd()`, having the special case handled there correctly all the time would prevent this accidental bug in most circumstances (the only manual parsing of `/etc/machine-id` is in `src/core/main.c` and `src/shared/dissect-image.c` which are both patched in this PR).

So I'd suggest keeping it like I implemented it right now.  I'd argue that it wouldn't really hurt other cases, as calling code already has to be robust against errors and whether `-EINVAL` or `-ENOMEDIUM` is returned in this one case shouldn't matter much.

Reply:
the thing is, we want in 99% of the cases that reading "uninitialized" from that file to be considered an error. After all we give the guarantee that /etc/machine-id is readable and initialized whenever any non-PID1 userspace is running. Thus, the only program that should ever see /etc/machine-id in this special state is PID 1 itself (and maybe systemd-machine-id-commit), all other tools should rely on it to be fully initiailzed

Reply:
I fully agree with you that any tool/software running _in_ a system should be able to expect that `/etc/machine-id` is fully initialized.

However tools conceptually running _outside_ a system, that is tools operating on files in a system-image that is currently not running (let's call this a cold image) must be robust against finding `$ROOT/etc/machine-id` in the `uninitialized` state.  I've gone through all the relevant call-sites now to see where this could pose an issue.  Here's a summary:

### `src/core/machine-id-setup.c` - `machine_id_setup()`
:heavy_check_mark: Just needs to check whether there is a valid machine-id.  Returning any error here means a machine-id needs to be set up, whether that's `ENOENT`, `ENOMEDIUM`, `EINVAL` does not matter.

### `src/libsystemd/sd-id128/sd-id128.c` - `sd_id128_get_machine()`
:heavy_check_mark: This is meant to be called from within a running system to retrieve the machine-id.  Thus this will always see it in an initialized state.

### `src/machine-id-setup/machine-id-setup-main.c` - `run()`
:heavy_check_mark: Only relevant in the `if (arg_commit)` code path which will only run from within a system.

### `src/nspawn/nspawn.c` - `setup_machine_id()`
:x: nspawn reads `/etc/machine-id` of the cold image it is attempting to spawn.  Here we need to be prepared against finding it in `uninitialized` state, especially because nspawn makes a distinction between 'missing machine-id' (is okay) and 'invalid machine-id' (will refuse to spawn).  I've added commit 572e347c7c9f46786ffc0a747f13df4598725f62 to address this.

### `src/partition/repart.c` - `context_read_seed()`
:x: As far as I understood, repart usually runs from initramfs and reads out the machine-id of the cold, not-yet-booted rootfs.  Thus this is also a situation where `uninitialized` must be dealt with properly.  Fixed with commit 378c41ca4ae768b40d562ecaa44b3afac230566d.

### `src/shared/machine-image.c` - `image_read_metadata()`
:heavy_check_mark: While this function seems to operate on a cold image as far as I can tell, it does not discriminate between between errors returned by `id128_read_fd()`.  Thus I would expect this code is only called when the image is supposed to be in good shape already?  Or is this a potentially not yet updated place that does not properly handle empty (and now `uninitialized`) machine-id files?

### `src/test/test-fs-util.c` - `test_chase_symlinks()`
:heavy_check_mark: Just reads the machine-id from the currently running system.

---

Honestly, I still don't think this change is a good idea.  It is really easy for a future reader of a "cold machine-id" to mess this up without knowing about it.  Just searching and understanding all the call-sites just now was a cumbersome task and I'd not be 100% confident that I catched everything.

The already covered situation of `/etc/machine-id` being an empty file will return `-ENOMEDIUM` in all situations right now (and most code is set up to deal with this correctly).  As the `uninitialized` value is conceptually very similar, I do not really see why it should not be treated equally and has to require this additional explicitness at call-site (which is easy to forget/don't know about).

That said, I've now implemented the change as requested if you prefer it this way.